src/osg/Texture.cpp

```diff
@@ -285,61 +315,102 @@ void Texture::TextureObject::setAllocated(GLint     numMipmapLevels,
+
+        case GL_COMPRESSED_RGB8_ETC2:                      numBitsPerTexel = 4.0;  break;
+        case GL_COMPRESSED_SRGB8_ETC2:                     numBitsPerTexel = 4.0;  break;
+        case GL_COMPRESSED_RGB8_PUNCHTHROUGH_ALPHA1_ETC2:  numBitsPerTexel = 8.0;  break;
+        case GL_COMPRESSED_SRGB8_PUNCHTHROUGH_ALPHA1_ETC2: numBitsPerTexel = 8.0;  break;
+        case GL_COMPRESSED_RGBA8_ETC2_EAC:                 numBitsPerTexel = 8.0;  break;
+        case GL_COMPRESSED_SRGB8_ALPHA8_ETC2_EAC:          numBitsPerTexel = 8.0;  break;
+        case GL_COMPRESSED_R11_EAC:                        numBitsPerTexel = 4.0;  break;
+        case GL_COMPRESSED_SIGNED_R11_EAC:                 numBitsPerTexel = 4.0;  break;
+        case GL_COMPRESSED_RG11_EAC:                       numBitsPerTexel = 8.0;  break;
+        case GL_COMPRESSED_SIGNED_RG11_EAC:                numBitsPerTexel = 8.0;  break;
+
+        // ASTC
+        case GL_COMPRESSED_RGBA_ASTC_4x4_KHR:
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_4x4_KHR:      numBitsPerTexel = 8.0;  break;
+        case GL_COMPRESSED_RGBA_ASTC_5x4_KHR:
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_5x4_KHR:      numBitsPerTexel = 6.4;  break;
+        case GL_COMPRESSED_RGBA_ASTC_5x5_KHR:
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_5x5_KHR:      numBitsPerTexel = 5.12; break;
+        case GL_COMPRESSED_RGBA_ASTC_6x5_KHR:
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_6x5_KHR:      numBitsPerTexel = 4.27; break;
+        case GL_COMPRESSED_RGBA_ASTC_6x6_KHR:
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_6x6_KHR:      numBitsPerTexel = 3.56; break;
+        case GL_COMPRESSED_RGBA_ASTC_8x5_KHR:
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_8x5_KHR:      numBitsPerTexel = 3.2;  break;
+        case GL_COMPRESSED_RGBA_ASTC_8x6_KHR:
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_8x6_KHR:      numBitsPerTexel = 2.67; break;
+        case GL_COMPRESSED_RGBA_ASTC_8x8_KHR:
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_8x8_KHR:      numBitsPerTexel = 2.0;  break;
+        case GL_COMPRESSED_RGBA_ASTC_10x5_KHR:
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x5_KHR:     numBitsPerTexel = 2.56; break;
+        case GL_COMPRESSED_RGBA_ASTC_10x6_KHR:
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x6_KHR:     numBitsPerTexel = 2.13; break;
+        case GL_COMPRESSED_RGBA_ASTC_10x8_KHR:
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x8_KHR:     numBitsPerTexel = 1.6;  break;
+        case GL_COMPRESSED_RGBA_ASTC_10x10_KHR:
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x10_KHR:    numBitsPerTexel = 1.28; break;
+        case GL_COMPRESSED_RGBA_ASTC_12x10_KHR:
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_12x10_KHR:    numBitsPerTexel = 1.07; break;
+        case GL_COMPRESSED_RGBA_ASTC_12x12_KHR:
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_12x12_KHR:    numBitsPerTexel = 0.89; break;
     }
 
     _size = (unsigned int)(ceil(double(_width * _height * _depth * numBitsPerTexel)/8.0));
```

Comment:
Do you mean, _size calculation is wrong here for ASTC format?

Reply:
I just explain what is wrong above.  The maths used doesn't work for block formats.  The correct way would be to compute the number of blocks in width and height separately, then multiple up to compute the final size.

Reply:
I understand you.
What about this implementation:

```
        // ASTC
        case GL_COMPRESSED_RGBA_ASTC_4x4_KHR:
        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_4x4_KHR:
        case GL_COMPRESSED_RGBA_ASTC_5x4_KHR:
        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_5x4_KHR:
        case GL_COMPRESSED_RGBA_ASTC_5x5_KHR:
        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_5x5_KHR:
        case GL_COMPRESSED_RGBA_ASTC_6x5_KHR:
        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_6x5_KHR:
        case GL_COMPRESSED_RGBA_ASTC_6x6_KHR:
        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_6x6_KHR:
        case GL_COMPRESSED_RGBA_ASTC_8x5_KHR:
        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_8x5_KHR:
        case GL_COMPRESSED_RGBA_ASTC_8x6_KHR:
        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_8x6_KHR:
        case GL_COMPRESSED_RGBA_ASTC_8x8_KHR:
        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_8x8_KHR:
        case GL_COMPRESSED_RGBA_ASTC_10x5_KHR:
        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x5_KHR:
        case GL_COMPRESSED_RGBA_ASTC_10x6_KHR:
        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x6_KHR:
        case GL_COMPRESSED_RGBA_ASTC_10x8_KHR:
        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x8_KHR:
        case GL_COMPRESSED_RGBA_ASTC_10x10_KHR:
        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x10_KHR:
        case GL_COMPRESSED_RGBA_ASTC_12x10_KHR:
        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_12x10_KHR:
        case GL_COMPRESSED_RGBA_ASTC_12x12_KHR:
        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_12x12_KHR:
        {
            _size = 0;
            for (int i = 0; i < std::max(_numMipmapLevels, 1); ++i)
            {
                GLint blockSize;
                GLint size;
                getCompressedSize(_internalFormat, _width/(1 << i), _height/(1 << i), _depth, blockSize, size);
                _size += size;
            }
            return;
        }
```

Reply:
Looking at Texture::getCompressedSize(..) it looks broken when handling the ASTC formats as well. The code looks like it could all do with a refactor to properly handling block formats of different block dimensions and total block memory size.

Reply:
FYI, with the VSG I implemented the compressed blocks as types much like vec3, vec4 etc, and have the image data stored in a Array,Array2D/Array3D.  The width, height are actually done in number of blocks, the final width passed to Vulkan for the image is the computed from the block size multiplied the block dimensions.  This avoids any problems with potentially data sizes being off.  There is none of this complex switch/if/else handling of block formats.

Alas trying to retrofit something like this to the OSG would required lots of changes to how osg::Image is managed so is not a practical solution.

It might gives us pause though - perhaps the OSG scheme is just broken.  I worked in adhoc way for the STC orginal formats but simply doesn't scale to handling all these new formats.  Trying to get it to scale is possibly the wrong approach, it perhaps should just be generalized in a better way to handle the different block formats.

What this might be I can't say.  I need to get back to VSG work now so will wrap up.